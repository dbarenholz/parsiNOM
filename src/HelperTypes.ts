import { Parser, STypeBase } from './Parser';

export type TupleToUnion<T extends readonly unknown[]> = T[number];

export type DeParser<T extends Parser<unknown>> = T extends Parser<infer U> ? U : never;

export type DeParserArray<T extends readonly Parser<unknown>[]> = {
	[I in keyof T]: T[I] extends Parser<infer R> ? R : never;
};
