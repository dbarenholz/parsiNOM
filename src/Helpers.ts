import { ParseResult, STypeBase } from './HelperTypes';
import { Parser } from './Parser';

export function arrayUnion(a: string[] | never[], b: string[] | never[]): string[] {
	const ret: string[] = [...a];
	for (const bElement of b) {
		let alreadyIncluded = false;
		for (const retElement of ret) {
			if (bElement === retElement) {
				alreadyIncluded = true;
			}
		}
		if (!alreadyIncluded) {
			ret.push(bElement);
		}
	}
	ret.sort();
	return ret;
}

// Parser Helpers

export class ParserHelpers {
	followedBy<SType extends STypeBase>(x: Parser<SType>): Parser<SType> {
		return new Parser<SType>(context => {
			const result = x.p(context.copy());
			result.position = context.position;
			return result;
		});
	}

	/**
	 * Returns a parser that yields undefined if the input parser fails, and fails if the input parser accepts. Consumes no input.
	 * More or less an inverse lookahead.
	 *
	 * @param parser
	 */
	notFollowedBy(parser: Parser<unknown>): Parser<undefined> {
		return new Parser((context): ParseResult<undefined> => {
			const contextCopy = context.copy();
			const result = parser.p(contextCopy);
			const text = context.sliceTo(contextCopy.position.index);
			return result.success ? context.fail(`not '` + text + `'`) : context.succeed(undefined);
		});
	}

	/**
	 * Returns a parser that passes the next character into `fn` and yields the character if `fn` returns true.
	 *
	 * @param fn
	 */
	test(fn: (char: string) => boolean): Parser<string> {
		return new Parser<string>((context): ParseResult<string> => {
			const char = context.input[context.position.index];
			if (!context.atEOF() && fn(char)) {
				return context.succeedOffset(1, char);
			} else {
				return context.fail(`a character matching ${fn}`);
			}
		});
	}
}

export const P_HELPERS = new ParserHelpers();
