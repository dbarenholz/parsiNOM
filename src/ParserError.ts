import { ParseFailure } from './HelperTypes';

function generateMessage(str: string, parseFailure: ParseFailure, verbose: boolean): string {
	const expectedMessage = `Expected ${parseFailure.expected.join(' or ')}`;
	let message = `Parse Failure: ${expectedMessage} at index ${parseFailure.furthest.index}, line ${parseFailure.furthest.line}, column ${parseFailure.furthest.column}.`;

	if (verbose) {
		const lines = str.split('\n');
		const failedLine = lines[parseFailure.furthest.line - 1]; // line is a one based index

		const linePrefix = `${parseFailure.furthest.line} |   `;
		message += `\n\n${linePrefix}${failedLine}`;
		message += `\n${' '.repeat(parseFailure.furthest.column - 1 + linePrefix.length)}^ (${expectedMessage})`;
	}

	return message;
}

export class ParsingError extends Error {
	constructor(str: string, parseFailure: ParseFailure) {
		super(generateMessage(str, parseFailure, true));
	}
}
