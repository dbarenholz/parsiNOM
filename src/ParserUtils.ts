import { Parser } from './Parser';
import { P } from './ParsiNOM';
import { ParsingPosition } from './HelperTypes';

export class P_UTILS {
	/**
	 * Yields the current position.
	 */
	static position(): Parser<ParsingPosition> {
		return new Parser<ParsingPosition>(context => {
			return context.succeed(context.position);
		});
	}

	/**
	 * Accepts any single character except for eof.
	 * Yields the character
	 */
	static any(): Parser<string> {
		return new Parser<string>(context => {
			if (context.atEOF()) {
				return context.fail('any character');
			}
			return context.succeedOffset(1, context.input[context.position.index]);
		});
	}

	/**
	 * Accepts the entire rest of the string until the end.
	 * Yields the rest of the string.
	 */
	static remaining(): Parser<string> {
		return new Parser<string>(context => {
			return context.succeedAt(context.input.length, context.input.slice(context.position.index));
		});
	}

	static eof(): Parser<undefined> {
		return new Parser<undefined>(context => {
			if (!context.atEOF()) {
				return context.fail('eof');
			}
			return context.succeed(undefined);
		});
	}

	static digit() {
		return P.regexp(/^[0-9]/).describe('a digit');
	}
	static digits() {
		return P.regexp(/^[0-9]+/).describe('optional digits');
	}
	static letter() {
		return P.regexp(/^[a-z]/i).describe('a letter');
	}
	static letters() {
		return P.regexp(/^[a-z]+/i).describe('optional letters');
	}
	static optionalWhitespace() {
		return P.regexp(/^\s*/).describe('optional whitespace');
	}
	static whitespace() {
		return P.regexp(/^\s+/).describe('whitespace');
	}
	static cr() {
		return P.string('\r');
	}
	static lf() {
		return P.string('\n');
	}
	static crlf() {
		return P.string('\r\n');
	}
	static newline() {
		return P.or(this.crlf(), this.lf(), this.cr()).describe('newline');
	}

	static prefix<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		nextParser: Parser<OtherSType>,
		combine: (a: OperatorSType, b: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		const parser: Parser<OtherSType | ReturnSType> = P.reference(() => {
			return P.sequenceMap(combine, operatorsParser, parser).or(nextParser);
		});
		return parser;
	}

	static postfix<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		nextParser: Parser<OtherSType>,
		combine: (a: OperatorSType, b: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(x, postfixes) => {
				return postfixes.reduce<OtherSType | ReturnSType>((acc, y) => combine(y, acc), x);
			},
			nextParser,
			operatorsParser.many(),
		);
	}

	static binaryRight<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		nextParser: Parser<OtherSType>,
		combine: (a: OtherSType, b: OperatorSType, c: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		const parser: Parser<OtherSType | ReturnSType> = P.reference(() =>
			P.sequenceMap(combine, nextParser, operatorsParser.trim(this.optionalWhitespace()), parser).or(nextParser),
		);
		return parser;
	}

	static binaryLeft<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		nextParser: Parser<OtherSType>,
		combine: (a: OtherSType | ReturnSType, b: OperatorSType, c: OtherSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(first: OtherSType, others) => {
				return others.reduce<OtherSType | ReturnSType>((acc, y) => {
					const [operator, next] = y;
					return combine(acc, operator, next);
				}, first);
			},
			nextParser,
			P.sequence(operatorsParser.trim(this.optionalWhitespace()), nextParser).many(),
		);
	}

	static func<ArgsSType, ReturnSType>(
		name: string | Parser<string>,
		args: Parser<ArgsSType>,
		combine: (name: string, args: ArgsSType) => ReturnSType,
	): Parser<ReturnSType> {
		const nameParser = typeof name === 'string' ? P.string(name) : name;
		return P.sequenceMap(
			(name, _lBracket, _1, args, _2, _rBracket) => {
				return combine(name, args);
			},
			nameParser,
			P.string('('),
			this.optionalWhitespace(),
			args,
			this.optionalWhitespace(),
			P.string(')'),
		);
	}
}
