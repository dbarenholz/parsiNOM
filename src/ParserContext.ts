import { arrayUnion } from './Helpers';
import { ParseResult, ParsingPosition, STypeBase } from './HelperTypes';

export class ParserContext {
	readonly input: string;
	position: ParsingPosition;

	constructor(input: string, position: ParsingPosition) {
		this.input = input;
		this.position = position;
	}

	moveToPosition(position: ParsingPosition): ParserContext {
		this.position = position;
		return this;
	}

	copy(): ParserContext {
		return new ParserContext(this.input, {
			index: this.position.index,
			column: this.position.column,
			line: this.position.line,
		});
	}

	atEOF(): boolean {
		return this.position.index >= this.input.length;
	}

	// OPTIMIZATION: only create a new position when needed, otherwise mutate
	private advanceTo(index: number): ParsingPosition {
		if (index < this.position.index) {
			throw new Error(`Can not advance backwards. Current pos ${this.position.index}. Advance target index ${index}.`);
		}
		if (index === this.position.index) {
			return this.position;
		}

		let line = this.position.line;
		let column = this.position.column;

		for (let i = this.position.index; i < index; i++) {
			if (this.input[i] === '\n') {
				line += 1;
				column = 1;
			} else {
				column += 1;
			}
		}

		this.position = {
			index: index,
			line: line,
			column: column,
		};

		return this.position;
	}

	private invalidPosition(): ParsingPosition {
		return {
			index: -1,
			line: -1,
			column: -1,
		};
	}

	sliceTo(endIndex: number): string {
		return this.input.slice(this.position.index, endIndex);
	}

	succeedOffset<SType extends STypeBase>(offset: number, value: SType): ParseResult<SType> {
		return this.succeedAt(this.position.index + offset, value);
	}

	failOffset<SType extends STypeBase>(offset: number, expected: string | string[]): ParseResult<SType> {
		return this.failAt(this.position.index + offset, expected);
	}

	succeed<SType extends STypeBase>(value: SType): ParseResult<SType> {
		return this.succeedAt(this.position.index, value);
	}

	fail<SType extends STypeBase>(expected: string | string[]): ParseResult<SType> {
		return this.failAt(this.position.index, expected);
	}

	succeedAt<SType extends STypeBase>(index: number, value: SType): ParseResult<SType> {
		return {
			success: true,
			position: this.advanceTo(index),
			value: value,
			furthest: this.invalidPosition(),
			expected: [],
		};
	}

	failAt<SType extends STypeBase>(index: number, expected: string | string[]): ParseResult<SType> {
		return {
			success: false,
			position: this.invalidPosition(),
			value: null,
			furthest: this.advanceTo(index),
			expected: Array.isArray(expected) ? expected : [expected],
		};
	}

	merge<ASType extends STypeBase, BSType extends STypeBase>(a: ParseResult<ASType> | undefined, b: ParseResult<BSType>): ParseResult<BSType> {
		if (a === undefined) {
			return b;
		}

		if (b.furthest.index > a.furthest.index) {
			return b;
		}

		const expected: string[] = b.furthest.index === a.furthest.index ? arrayUnion(a.expected, b.expected) : a.expected;

		b.furthest = a.furthest;
		b.expected = expected;
		return b;
	}
}
