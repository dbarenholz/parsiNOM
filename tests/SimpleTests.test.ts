import { P } from '../src/ParsiNOM';
import { testParse } from './TestHelpers';
import { P_UTILS } from '../src/ParserUtils';

describe('single string', () => {
	const parser = P.string('aba').thenEof();
	const matchingTable: [string, boolean][] = [
		['aba', true],
		['a', false],
		['ab', false],
		['abab', false],
		['ababa', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('many', () => {
	const parser = P.string('a').many().thenEof();
	const matchingTable: [string, boolean][] = [
		['a', true],
		['aa', true],
		['aaa', true],
		['abaa', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('sequence', () => {
	const parser = P.sequence(P.string('a'), P.string('b'));
	const matchingTable: [string, boolean][] = [
		['ab', true],
		['aab', false],
		['a', false],
		['aa', false],
		['baba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('sequence eof', () => {
	const parser = P.sequence(P.string('a'), P.string('b')).thenEof();
	const matchingTable: [string, boolean][] = [
		['ab', true],
		['aab', false],
		['a', false],
		['aa', false],
		['baba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('ab sequence many', () => {
	const parser = P.sequence(P.string('a'), P.string('b')).many().thenEof();
	const matchingTable: [string, boolean][] = [
		['ab', true],
		['abab', true],
		['a', false],
		['aa', false],
		['aba', false],
		['baba', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('binary left', () => {
	const parser = P_UTILS.binaryLeft(P.string('+'), P.string('a'), (a, b, c) => [a, b, c]).thenEof();
	const matchingTable: [string, boolean][] = [
		['a+a', true],
		['a + a', true],
		['a', true],
		['a + b', false],
		['ab+', false],
		['+ab', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('binary right', () => {
	const parser = P_UTILS.binaryRight(P.string('+'), P.string('a'), (a, b, c) => [a, b, c]).thenEof();
	const matchingTable: [string, boolean][] = [
		['a+a', true],
		['a + a', true],
		['a', true],
		['a + b', false],
		['ab+', false],
		['+ab', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('prefix', () => {
	const parser = P_UTILS.prefix(P.string('-'), P.string('a'), (a, b) => [a, b]).thenEof();
	const matchingTable: [string, boolean][] = [
		['a', true],
		['-a', true],
		['--a', true],
		['a - b', false],
		['ab-', false],
		['-ab', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('postfix', () => {
	const parser = P_UTILS.postfix(P.string('-'), P.string('a'), (a, b) => [a, b]).thenEof();
	const matchingTable: [string, boolean][] = [
		['a', true],
		['a-', true],
		['a--', true],
		['a- b', false],
		['ab-', false],
		['-ab', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('trim', () => {
	const parser = P.string('a').trim(P_UTILS.optWhitespace()).thenEof();
	const matchingTable: [string, boolean][] = [
		['a', true],
		['a ', true],
		['a  ', true],
		[' a', true],
		['  a ', true],
		[' ab', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('sequence trim', () => {
	const parser = P.sequence(P.string('b'), P.string('a').trim(P_UTILS.optWhitespace()), P.string('b')).thenEof();
	const matchingTable: [string, boolean][] = [
		['bab', true],
		['ba b', true],
		['ba  b', true],
		['b ab', true],
		['b  a b', true],
		['b abb', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('lazy sequence trim', () => {
	const parser = P.reference(() => P.sequence(P.string('b'), P.string('a').trim(P_UTILS.optWhitespace()), P.string('b')).thenEof());
	const matchingTable: [string, boolean][] = [
		['bab', true],
		['ba b', true],
		['ba  b', true],
		['b ab', true],
		['b  a b', true],
		['b abb', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('or', () => {
	const parser = P.or(P.string('a'), P.string('b'), P.string('c')).thenEof();
	const matchingTable: [string, boolean][] = [
		['a', true],
		['b', true],
		['c', true],
		['bb', false],
		['b  a b', false],
		['b abb', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('or sequence', () => {
	const parser = P.sequence(P.or(P.string('ab'), P.string('c')), P.string('de'));
	const matchingTable: [string, boolean][] = [
		['abde', true],
		['cde', true],
		['c', false],
		['ab', false],
		['aba', false],
		['abc', false],
		['bcaba', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('regex', () => {
	const parser = P.regexp(/[0-9]+/)
		.map(str => Number.parseInt(str))
		.describe('number')
		.mark()
		.thenEof();
	const matchingTable: [string, boolean][] = [
		['1', true],
		['12', true],
		['23', true],
		['4234', true],
		['234a', false],
		['a123', false],
		['a1a', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});
