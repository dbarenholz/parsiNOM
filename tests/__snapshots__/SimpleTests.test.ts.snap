// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ab sequence many 'a' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`ab sequence many 'aa' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`ab sequence many 'ab' AST to match snapshot 1`] = `
[
  [
    "a",
    "b",
  ],
]
`;

exports[`ab sequence many 'aba' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 4,
    "index": 3,
    "line": 1,
  },
}
`;

exports[`ab sequence many 'abab' AST to match snapshot 1`] = `
[
  [
    "a",
    "b",
  ],
  [
    "a",
    "b",
  ],
]
`;

exports[`ab sequence many 'baba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
    "eof",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`ab sequence many 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
    "eof",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`binary left '+ab' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`binary left 'a + a' AST to match snapshot 1`] = `
[
  "a",
  "+",
  "a",
]
`;

exports[`binary left 'a + b' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 5,
    "index": 4,
    "line": 1,
  },
}
`;

exports[`binary left 'a' AST to match snapshot 1`] = `"a"`;

exports[`binary left 'a+a' AST to match snapshot 1`] = `
[
  "a",
  "+",
  "a",
]
`;

exports[`binary left 'ab+' Error to match snapshot 1`] = `
{
  "expected": [
    "'+'",
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`binary left 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`binary right '+ab' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`binary right 'a + a' AST to match snapshot 1`] = `
[
  "a",
  "+",
  "a",
]
`;

exports[`binary right 'a + b' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 5,
    "index": 4,
    "line": 1,
  },
}
`;

exports[`binary right 'a' AST to match snapshot 1`] = `"a"`;

exports[`binary right 'a+a' AST to match snapshot 1`] = `
[
  "a",
  "+",
  "a",
]
`;

exports[`binary right 'ab+' Error to match snapshot 1`] = `
{
  "expected": [
    "'+'",
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`binary right 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`lazy sequence trim 'b  a b' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`lazy sequence trim 'b ab' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`lazy sequence trim 'b abb' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 5,
    "index": 4,
    "line": 1,
  },
}
`;

exports[`lazy sequence trim 'ba  b' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`lazy sequence trim 'ba b' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`lazy sequence trim 'bab' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`lazy sequence trim 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`many 'a' AST to match snapshot 1`] = `
[
  "a",
]
`;

exports[`many 'aa' AST to match snapshot 1`] = `
[
  "a",
  "a",
]
`;

exports[`many 'aaa' AST to match snapshot 1`] = `
[
  "a",
  "a",
  "a",
]
`;

exports[`many 'abaa' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`or 'a' AST to match snapshot 1`] = `"a"`;

exports[`or 'b  a b' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`or 'b abb' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`or 'b' AST to match snapshot 1`] = `"b"`;

exports[`or 'bb' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`or 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`or 'c' AST to match snapshot 1`] = `"c"`;

exports[`or sequence 'ab' Error to match snapshot 1`] = `
{
  "expected": [
    "'de'",
  ],
  "pos": {
    "column": 3,
    "index": 2,
    "line": 1,
  },
}
`;

exports[`or sequence 'aba' Error to match snapshot 1`] = `
{
  "expected": [
    "'de'",
  ],
  "pos": {
    "column": 3,
    "index": 2,
    "line": 1,
  },
}
`;

exports[`or sequence 'abc' Error to match snapshot 1`] = `
{
  "expected": [
    "'de'",
  ],
  "pos": {
    "column": 3,
    "index": 2,
    "line": 1,
  },
}
`;

exports[`or sequence 'abde' AST to match snapshot 1`] = `
[
  "ab",
  "de",
]
`;

exports[`or sequence 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'ab'",
    "'c'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`or sequence 'c' Error to match snapshot 1`] = `
{
  "expected": [
    "'de'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`or sequence 'cde' AST to match snapshot 1`] = `
[
  "c",
  "de",
]
`;

exports[`postfix '-ab' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`postfix 'a' AST to match snapshot 1`] = `"a"`;

exports[`postfix 'a- b' Error to match snapshot 1`] = `
{
  "expected": [
    "'-'",
    "eof",
  ],
  "pos": {
    "column": 3,
    "index": 2,
    "line": 1,
  },
}
`;

exports[`postfix 'a-' AST to match snapshot 1`] = `
[
  "-",
  "a",
]
`;

exports[`postfix 'a--' AST to match snapshot 1`] = `
[
  "-",
  [
    "-",
    "a",
  ],
]
`;

exports[`postfix 'ab-' Error to match snapshot 1`] = `
{
  "expected": [
    "'-'",
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`postfix 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`prefix '--a' AST to match snapshot 1`] = `
[
  "-",
  [
    "-",
    "a",
  ],
]
`;

exports[`prefix '-a' AST to match snapshot 1`] = `
[
  "-",
  "a",
]
`;

exports[`prefix '-ab' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 3,
    "index": 2,
    "line": 1,
  },
}
`;

exports[`prefix 'a - b' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`prefix 'a' AST to match snapshot 1`] = `"a"`;

exports[`prefix 'ab-' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`prefix 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'-'",
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`regex '1' AST to match snapshot 1`] = `
{
  "range": {
    "from": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "to": {
      "column": 2,
      "index": 1,
      "line": 1,
    },
  },
  "value": 1,
}
`;

exports[`regex '12' AST to match snapshot 1`] = `
{
  "range": {
    "from": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "to": {
      "column": 3,
      "index": 2,
      "line": 1,
    },
  },
  "value": 12,
}
`;

exports[`regex '23' AST to match snapshot 1`] = `
{
  "range": {
    "from": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "to": {
      "column": 3,
      "index": 2,
      "line": 1,
    },
  },
  "value": 23,
}
`;

exports[`regex '234a' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 4,
    "index": 3,
    "line": 1,
  },
}
`;

exports[`regex '4234' AST to match snapshot 1`] = `
{
  "range": {
    "from": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "to": {
      "column": 5,
      "index": 4,
      "line": 1,
    },
  },
  "value": 4234,
}
`;

exports[`regex 'a1a' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`regex 'a123' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 4,
    "index": 3,
    "line": 1,
  },
}
`;

exports[`sequence 'a' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`sequence 'aa' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`sequence 'aab' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`sequence 'ab' AST to match snapshot 1`] = `
[
  "a",
  "b",
]
`;

exports[`sequence 'baba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`sequence eof 'a' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`sequence eof 'aa' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`sequence eof 'aab' Error to match snapshot 1`] = `
{
  "expected": [
    "'b'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`sequence eof 'ab' AST to match snapshot 1`] = `
[
  "a",
  "b",
]
`;

exports[`sequence eof 'baba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`sequence trim 'b  a b' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`sequence trim 'b ab' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`sequence trim 'b abb' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 5,
    "index": 4,
    "line": 1,
  },
}
`;

exports[`sequence trim 'ba  b' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`sequence trim 'ba b' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`sequence trim 'bab' AST to match snapshot 1`] = `
[
  "b",
  "a",
  "b",
]
`;

exports[`sequence trim 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 2,
    "index": 1,
    "line": 1,
  },
}
`;

exports[`single string 'a' Error to match snapshot 1`] = `
{
  "expected": [
    "'aba'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`single string 'ab' Error to match snapshot 1`] = `
{
  "expected": [
    "'aba'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;

exports[`single string 'aba' AST to match snapshot 1`] = `"aba"`;

exports[`single string 'abab' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 4,
    "index": 3,
    "line": 1,
  },
}
`;

exports[`single string 'ababa' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 4,
    "index": 3,
    "line": 1,
  },
}
`;

exports[`trim '  a ' AST to match snapshot 1`] = `"a"`;

exports[`trim ' a' AST to match snapshot 1`] = `"a"`;

exports[`trim ' ab' Error to match snapshot 1`] = `
{
  "expected": [
    "eof",
  ],
  "pos": {
    "column": 3,
    "index": 2,
    "line": 1,
  },
}
`;

exports[`trim 'a  ' AST to match snapshot 1`] = `"a"`;

exports[`trim 'a ' AST to match snapshot 1`] = `"a"`;

exports[`trim 'a' AST to match snapshot 1`] = `"a"`;

exports[`trim 'bcaba' Error to match snapshot 1`] = `
{
  "expected": [
    "'a'",
  ],
  "pos": {
    "column": 1,
    "index": 0,
    "line": 1,
  },
}
`;
